cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,805 @@
+          [key_func =
+             cugraph::detail::compute_gpu_id_from_ext_edge_t<edge_t>{
+               comm_size,
+               major_comm_size,
+               minor_comm_size}] __device__(auto,
+                                            auto,
+                                            thrust::nullopt_t,
+                                            thrust::nullopt_t,
+                                            thrust::tuple<edge_t, edge_type_t> id_and_type) {
+            return thrust::optional<thrust::tuple<int, edge_type_t>>{thrust::make_tuple(
+              key_func(thrust::get<0>(id_and_type)), thrust::get<1>(id_and_type))};
+          }));
+
+    auto type_and_gpu_id_pair_begin =
+      thrust::make_zip_iterator(thrust::make_tuple(edge_types.begin(), gpu_ids.begin()));
+
+    thrust::sort(handle.get_thrust_policy(),
+                 type_and_gpu_id_pair_begin,
+                 type_and_gpu_id_pair_begin + edge_types.size());
+
+    auto nr_unique_paris = thrust::count_if(
+      handle.get_thrust_policy(),
+      thrust::make_counting_iterator(size_t{0}),
+      thrust::make_counting_iterator(edge_types.size()),
+      detail::is_first_in_run_t<decltype(type_and_gpu_id_pair_begin)>{type_and_gpu_id_pair_begin});
+
+    auto unique_pairs_buffer = cugraph::allocate_dataframe_buffer<
+      typename thrust::iterator_traits<decltype(type_and_gpu_id_pair_begin)>::value_type>(
+      nr_unique_paris, handle.get_stream());
+
+    rmm::device_uvector<edge_t> unique_pair_counts(nr_unique_paris, handle.get_stream());
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          type_and_gpu_id_pair_begin,
+                          type_and_gpu_id_pair_begin + edge_types.size(),
+                          thrust::make_constant_iterator(size_t{1}),
+                          cugraph::get_dataframe_buffer_begin(unique_pairs_buffer),
+                          unique_pair_counts.begin());
+
+    edge_types.resize(0, handle.get_stream());
+    gpu_ids.resize(0, handle.get_stream());
+
+    std::forward_as_tuple(
+      std::tie(
+        std::get<0>(unique_pairs_buffer), std::get<1>(unique_pairs_buffer), unique_pair_counts),
+      std::ignore) =
+      cugraph::groupby_gpu_id_and_shuffle_values(
+        handle.get_comms(),
+        thrust::make_zip_iterator(thrust::make_tuple(std::get<0>(unique_pairs_buffer).begin(),
+                                                     std::get<1>(unique_pairs_buffer).begin(),
+                                                     unique_pair_counts.begin())),
+        thrust::make_zip_iterator(thrust::make_tuple(std::get<0>(unique_pairs_buffer).end(),
+                                                     std::get<1>(unique_pairs_buffer).end(),
+                                                     unique_pair_counts.end())),
+        [] __device__(auto val) { return thrust::get<1>(val); },
+        handle.get_stream());
```

Comment:
If you sort with the gpu_id as the primary key and the edge type as the secondary key, this can be just shuffle_values, right?

Reply:
To use shuffle_values,  `(type, gpu_id, #element_for_this_type) ` tuples need  to be grouped by counted based on gpu_id and then we can call  shuffle_values to get  (type,  #element_for_this_type)  pairs to the right gpu.